### DISCUSSION TITLE
separate low mid and high frequencies in the editor waveform

### DISCUSSION THREAD

**Discussion by Doksha** [ppy/osu#20935] [Ideas]
The waveform is useful for timing and analyzing music that has distinct frequencies, but noisier music is basically just one big blob, with no discernable rhythm in the waveform. This could be solved by splitting the frequencies into low, mid, and high (separated by color, red being low, blue being mid, and green being high).

**Comment by peppy** [comment]
They already are. Feel free to tweak the colours / ranges and propose something which doesn't look like unicorn barf but conveys more detail.

**Reply by Doksha** [reply]
I was thinking of a software I had used a while ago and forgot which colors it used; it uses blue for lows, yellow for mids, and white for highs (mixing colors for overlapping frequencies.) it looks pretty good and conveys a lot more detail than the lazer waveform 

Also what do you mean they already are? there isnt anything noticable, and there arent any settings I can find related to waveform

**Reply by peppy** [reply]
I intentionally chose similar colours because using completely different colours is hard to make look good

> Also what do you mean they already are? there isn't anything noticeable, and there aren't any settings I can find related to waveform

You will need to change in the code, there's no settings.

**Reply by smoogipoo** [reply]
> Also what do you mean they already are? there isnt anything noticable, and there arent any settings I can find related to waveform

It's very faint, but it does have low/mid/high separation:

![image](https://user-images.githubusercontent.com/1329837/197919819-9ea4350d-524a-497e-9c5f-0baa63624a98.png)

Frequencies defined here:
https://github.com/ppy/osu-framework/blob/88e3ad7ae52da410e139a041d03c696508b88f0b/osu.Framework/Audio/Track/Waveform.cs#L43-L61

And colours defined here:
https://github.com/ppy/osu/blob/181c7d1d6e9900225321f6170948bbe96767ac08/osu.Game/Screens/Edit/Compose/Components/Timeline/Timeline.cs#L121-L123

**Reply by Doksha** [reply]
![This](https://user-images.githubusercontent.com/72625258/197940330-90c34b92-11cf-4ad4-8a8c-8798a3607da3.png) This is what I mean by frequency separation, you can see where each beat is, without being drowned out by the bass.
I'm guessing that (smoogipoo's image) just takes the peak frequency, finds where it lands in high, mid, or low, then sets the color to that band, it still doesn't provide much actual information on where beats are in noisy music.

There needs to be separate waveforms for each of the frequency bands, current system is just an inconsistent tinting effect.

I understand now what peppy was talking about, and yes, with the current system, really any colors would look like "unicorn barf"
I noticed the slight tinting effect and just though that was an artifact of how the waveforms were rendered.

An easy alternative would be to simply boost the highs in all waveforms and ignore coloring entirely.

**Reply by peppy** [reply]
Yeah, I get that. I started to write up an issue to propose implementing that style, but on checking multiple apps I remember why I didn't do that in the first place. The visuals of it are pretty hard to tune, and it will make the drawing quite a lot more complex, but the larger reason is that apps like Traktor do a very good job with single colours for the full height (google "traktor waveform" or similar) and I think we can get similar levels of details without adding the visual clutter of separating out the frequency bands vertically.
